private/buf/buffetch/internal/reader.go

```diff
@@ -583,101 +748,156 @@ func findTerminateFileDirectoryPathFromBucket(
 		}
 		parent := normalpath.Dir(terminateFileDirectoryPath)
 		if parent == terminateFileDirectoryPath {
-			return "", nil
+			break
 		}
 		terminateFileDirectoryPath = parent
 	}
+	return newTerminateFilePriority(terminateFiles), nil
 }
 
-func anyExistsBucket(
+func terminateFilesInBucket(
 	ctx context.Context,
 	readBucket storage.ReadBucket,
-	paths []string,
-) (bool, error) {
-	for _, path := range paths {
-		exists, err := storage.Exists(ctx, readBucket, path)
-		if err != nil {
-			return false, err
-		}
-		if exists {
-			return true, nil
+	paths [][]string,
+) ([]TerminateFile, error) {
+	foundPaths := make([]TerminateFile, len(paths))
+	for i := range paths {
+		for _, path := range paths[i] {
+			exists, err := storage.Exists(ctx, readBucket, path)
+			if err != nil {
+				return nil, err
+			}
+			if exists {
+				foundPaths[i] = newTerminateFile(filepath.Base(path), normalpath.Normalize(filepath.Dir(path)))
+				// We want the first file found for each layer of hierarchy.
+				break
+			}
 		}
 	}
-	return false, nil
+	return foundPaths, nil
 }
 
 // findTerminateFileDirectoryPathFromOS returns the directory that contains
 // the terminateFileName, starting with the subDirPath and ascending until
-// the root of the local filesysem.
-func findTerminateFileDirectoryPathFromOS(subDirPath string, terminateFileNames []string) (string, error) {
+// the root of the local filesystem.
+func findTerminateFileDirectoryPathFromOS(
+	subDirPath string,
+	terminateFileNames [][]string,
+) (TerminateFilePriority, error) {
 	if len(terminateFileNames) == 0 {
-		return "", nil
+		return nil, nil
 	}
 	fileInfo, err := os.Stat(normalpath.Unnormalize(subDirPath))
 	if err != nil {
 		if os.IsNotExist(err) {
-			return "", storage.NewErrNotExist(subDirPath)
+			return nil, storage.NewErrNotExist(subDirPath)
 		}
-		return "", err
+		return nil, err
 	}
 	if !fileInfo.IsDir() {
-		return "", normalpath.NewError(normalpath.Unnormalize(subDirPath), errors.New("not a directory"))
+		return nil, normalpath.NewError(normalpath.Unnormalize(subDirPath), errors.New("not a directory"))
 	}
+	terminateFiles := make([]TerminateFile, len(terminateFileNames))
 	terminateFileDirectoryPath, err := normalpath.NormalizeAndAbsolute(subDirPath)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
+	foundFiles := 0
 	for {
-		fullTerminateFileNames := make([]string, len(terminateFileNames))
-		for i, terminateFileName := range terminateFileNames {
-			fullTerminateFileNames[i] = normalpath.Unnormalize(normalpath.Join(terminateFileDirectoryPath, terminateFileName))
+		fullTerminateFileNames := make([][]string, len(terminateFileNames))
+		for i := range terminateFileNames {
+			fullTerminateFileNames[i] = make([]string, len(terminateFileNames[i]))
+			for j, terminateFileName := range terminateFileNames[i] {
+				fullTerminateFileNames[i][j] = normalpath.Unnormalize(normalpath.Join(terminateFileDirectoryPath, terminateFileName))
+			}
```

Comment:
Similar to the other path transformation comment, we might be able to just pass in the current directory to `terminateFileOnOS` and un-normalize the path at the `os.Stat` call-site.